diff --git a/usr/src/Makefile.master b/usr/src/Makefile.master
index 9e85a2fa72..8c74d39d76 100644
--- a/usr/src/Makefile.master
+++ b/usr/src/Makefile.master
@@ -597,8 +597,8 @@ CPPFLAGS.native=$(ENVCPPFLAGS1) $(ENVCPPFLAGS2) $(ENVCPPFLAGS3) \
 		$(ENVCPPFLAGS4) -I$(NATIVE_ADJUNCT)/include
 CPPFLAGS=	$(CPPFLAGS.first) $(CPPFLAGS.master)
 AS_CPPFLAGS=	$(CPPFLAGS.first) $(CPPFLAGS.master)
-JAVAFLAGS=	-source 7 -target 7 -Xlint:deprecation,-options
-$(BLD_JAVA_11)JAVAFLAGS=	-source 7 -target 7 -Xlint:-options
+JAVAFLAGS=	-source 8 -target 8 -Xlint:deprecation,-options
+$(BLD_JAVA_11)JAVAFLAGS=	-source 8 -target 8 -Xlint:none
 
 #
 # For source message catalogue
diff --git a/usr/src/cmd/dtrace/test/cmd/jdtrace/Makefile b/usr/src/cmd/dtrace/test/cmd/jdtrace/Makefile
index 4d88a584de..1fa09813db 100644
--- a/usr/src/cmd/dtrace/test/cmd/jdtrace/Makefile
+++ b/usr/src/cmd/dtrace/test/cmd/jdtrace/Makefile
@@ -73,7 +73,7 @@ $(PROG): $(SRCS)
 	$(POST_PROCESS) ; $(STRIP_STABS)
 
 JFLAGS= -g -cp $(CLASSPATH) -d $(CLASSDIR)
-JFLAGS += -source 7 -target 7 -Xlint:all,-options
+JFLAGS += -source 8 -target 8 -Xlint:none
 COMPILE.java=$(JAVAC) $(JFLAGS)
 JAVASRC= JDTrace.java Getopt.java
 
diff --git a/usr/src/cmd/dtrace/test/tst/common/java_api/Makefile b/usr/src/cmd/dtrace/test/tst/common/java_api/Makefile
index 982e906cf0..9083ee16da 100644
--- a/usr/src/cmd/dtrace/test/tst/common/java_api/Makefile
+++ b/usr/src/cmd/dtrace/test/tst/common/java_api/Makefile
@@ -58,7 +58,7 @@ all: default
 install: all $(PROTO_TEST_JAR)
 
 JFLAGS= -g -cp $(CLASSPATH) -d $(CLASSDIR)
-JFLAGS += -source 7 -target 7 -Xlint:all,-deprecation,-options,-rawtypes
+JFLAGS += -source 8 -target 8 -Xlint:none
 COMPILE.java=$(JAVAC) $(JFLAGS)
 
 $(TEST_JAR): $(SRCDIR)/*.java
diff --git a/usr/src/cmd/pools/poold/Makefile b/usr/src/cmd/pools/poold/Makefile
index 9dfb547196..ea58e3f83a 100644
--- a/usr/src/cmd/pools/poold/Makefile
+++ b/usr/src/cmd/pools/poold/Makefile
@@ -141,7 +141,7 @@ $(SUBDIRS): FRC
 
 doc:	all
 	$(JAVADOC) -private -classpath $(JARFILE): -sourcepath . \
-	    -source 7 -d $(DOCDIR) $(PACKAGEDOCS)
+	    -source 8 -d $(DOCDIR) $(PACKAGEDOCS)
 
 $(PROG):	$(OBJS)
 	$(LINK.c) -o $(PROG) $(OBJS) $(LDLIBS)
diff --git a/usr/src/lib/libdtrace_jni/java/Makefile b/usr/src/lib/libdtrace_jni/java/Makefile
index e48b5501a4..614f9ee4e7 100644
--- a/usr/src/lib/libdtrace_jni/java/Makefile
+++ b/usr/src/lib/libdtrace_jni/java/Makefile
@@ -66,8 +66,8 @@ APICLASSPATH=$(CLASSES):$(APIJAVASRC)
 
 # javac flags
 JFLAGS= -g -d $(CLASSES) -sourcepath $(APICLASSPATH)
-JFLAGS += -source 7 -target 7
-JFLAGS += -Xlint:all,-deprecation,-rawtypes,-options
+JFLAGS += -source 8 -target 8
+JFLAGS += -Xlint:none
 
 # The default make rule for Java files
 COMPILE.java=$(JAVAC) $(JFLAGS)
