--- copyn-gate/usr/src/lib/fm/libfmd_snmp/Makefile.com	Tue Feb 18 20:36:26 2020
+++ rti-gate/usr/src/lib/fm/libfmd_snmp/Makefile.com	Tue Feb 18 08:18:50 2020
@@ -22,8 +22,6 @@
 # Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
 #
-# Copyright 2020 Joyent, Inc.
-#
 
 LIBRARY = libfmd_snmp.a
 VERS = .1
@@ -53,8 +51,8 @@
 CFLAGS += $(CCVERBOSE) $(C_BIGPICFLAGS)
 CFLAGS64 += $(CCVERBOSE) $(C_BIGPICFLAGS)
 
-SNMPLIBS = -lnetsnmp -lnetsnmpagent
-NATIVE_LIBS += libnetsnmp.so libnetsnmpagent.so
+SNMPLIBS = -lnetsnmp -lnetsnmphelpers -lnetsnmpagent
+NATIVE_LIBS += libnetsnmp.so libnetsnmphelpers.so libnetsnmpagent.so
 
 LDLIBS += $(MACH_LDLIBS)
 LDLIBS += -lfmd_adm -luutil -lnvpair -ltopo
