--- rti-gate/usr/src/lib/krb5/plugins/kdb/ldap/Makefile.com	Tue Feb 18 08:18:50 2020
+++ copyn-gate/usr/src/lib/krb5/plugins/kdb/ldap/Makefile.com	Tue Feb 18 22:06:56 2020
@@ -65,8 +65,7 @@
 
 DYNFLAGS +=	$(KERBRUNPATH)
 # setting -L $(ROOT)/usr/lib/gss because libkdb_ldap needs mech_krb5
-LDLIBS +=	-L $(ROOT)/usr/lib/gss -L $(ROOTLIBDIR) -lkdb_ldap \
-		-lc
+LDLIBS +=	-L $(ROOT)/usr/lib/gss -L $(ROOTLIBDIR) -lkdb_ldap
 
 .KEEP_STATE:
 
