diff --git a/usr/src/pkg/manifests/developer-debug-mdb.p5m b/usr/src/pkg/manifests/developer-debug-mdb.p5m
index 5614b7c27f..d398997fb3 100644
--- a/usr/src/pkg/manifests/developer-debug-mdb.p5m
+++ b/usr/src/pkg/manifests/developer-debug-mdb.p5m
@@ -158,8 +158,6 @@ file path=usr/lib/mdb/proc/$(ARCH64)/libavl.so group=sys mode=0555
 file path=usr/lib/mdb/proc/$(ARCH64)/libc.so group=sys mode=0555
 file path=usr/lib/mdb/proc/$(ARCH64)/libnvpair.so group=sys mode=0555
 file path=usr/lib/mdb/proc/$(ARCH64)/libproc.so group=sys mode=0555
-file path=usr/lib/mdb/proc/$(ARCH64)/libpython$(PYTHON3_VERSION).so group=sys \
-    mode=0555
 $(i386_ONLY)file path=usr/lib/mdb/proc/$(ARCH64)/libsmbios.so group=sys \
     mode=0555
 file path=usr/lib/mdb/proc/$(ARCH64)/libsysevent.so group=sys mode=0555
