diff --git a/usr/src/uts/sparc/Makefile.sparc b/usr/src/uts/sparc/Makefile.sparc
index 990ccea..a58373f 100644
--- a/usr/src/uts/sparc/Makefile.sparc
+++ b/usr/src/uts/sparc/Makefile.sparc
@@ -77,8 +77,6 @@ LINT64_FILES	 = $(LINT64_DIRS:%=%/llib-l$(MODULE).ln)
 #	interfaces.)
 #
 include $(UTSBASE)/sparc/Makefile.files
-include $(UTSBASE)/sparc/v9/Makefile.files
-include $(UTSBASE)/sun/Makefile.files
 include $(UTSBASE)/common/Makefile.files
 
 #
@@ -202,7 +200,7 @@ ALL_DEFS	 = $(MACHINE_DEFS) $(DEBUG_DEFS) $(OPTION_DEFS)
 #
 #	Common Drivers (usually pseudo drivers) (/kernel/drv):
 #
-DRV_KMODS	+= foo
+DRV_KMODS	+=
 
 #
 #	Exec Class Modules (/kernel/exec):
